File: trabalho-sistema-eleitoral-brasileiro-java/TP1/src/Candidate.java
package TP1;

import java.util.Calendar;
import java.util.Formatter;

public class Candidate extends Person implements Comparable<Candidate> {
    private String balBoxName;
    private int balBoxNumber;
    private int nominalVotes;
    private String situation;
    private int partyNumber;
    private String voteDestination;
    private PoliticParty party;

    public Candidate(String name, char gender, String birth, String balBoxName, int balBoxNumber, int nominalVotes, String situation, int partyNumber, String voteDestination) {
        super(name, gender, birth);
        
        this.balBoxName = balBoxName;
        this.balBoxNumber = balBoxNumber;
        this.nominalVotes = nominalVotes;
        this.situation = situation;
        this.partyNumber = partyNumber;
        this.voteDestination = voteDestination;
    }

	public String getBalBoxName() {
		return this.balBoxName;
	}

	public int getBalBoxNumber() {
		return this.balBoxNumber;
	}

	public int getNominalVotes() {
		return this.nominalVotes;
	}

	public String getSituation() {
		return this.situation;
	}

    public int getPartyNumber(){
        return this.partyNumber;
    }
    
    public PoliticParty getParty(){
        return this.party;
    }
    
    public void setParty(PoliticParty party) {
    	this.party = party;
    }

    public String getVoteDestination() {
		return this.voteDestination;
	}
    
    public Boolean elected() {
    	return this.getSituation().equals("Eleito");
    }
    
    public Boolean validVote() {
    	return this.getVoteDestination().equals("Válido");
    } 
    
    /**
 	 * @brief Método que auxilia a impressão de um candidato.
 	 * @return String que contém os campos de um candidato.
 	 */
    @Override
    public String toString() {
        Formatter fmt = new Formatter();
        fmt.format("Name: %s\nVoteDestination: %s\n", this.getName(), this.voteDestination);
        String s = fmt.out().toString();
        fmt.close();
        return s;
    }
    
    /**
  	 * @brief Esse método auxilia a ordenção dos candidatos. 
  	 * Nele, implementa-se a comparação por votos nominais entre dois candidatos.
  	 * Caso os votos nominais sejam iguam, o próximo critério analisado é a idade.
  	 * O candidato mais velho possui preferência sobre o mais novo.
  	 * @param:
 	 * 		@Candadidate Candidato que terá seus votos nominais comparados com os votos nominais de outro candidato.
  	 * @return Retorna um valor positivo ou negativo dependendo dos votos nominais dos candidatos comparados.
  	 */
    @Override
    public int compareTo(Candidate candidate) {
        if(candidate.nominalVotes == this.nominalVotes) {
            Calendar birth1 = this.getBirth();
            Calendar birth2 = candidate.getBirth();

            if(birth1.before(birth2)) return -1; // Candidate 1 is older 
            else return 1;   

        } else {
            return candidate.nominalVotes - this.nominalVotes;
        }
    }
}

File: trabalho-sistema-eleitoral-brasileiro-cpp/src/Candidate.cpp
#include "../include/Candidate.h"
// #include "Candidate.h"

Candidate::Candidate(const string& name, char gender, const string& birth, const string& balboxName, int balboxNumber, int nominalVotes, const string& situation, int partyNumber, const string& voteDestination): Person(name, gender, birth) {
    this->balboxName = balboxName;
    this->balboxNumber = balboxNumber;
    this->nominalVotes = nominalVotes;
    this->situation = situation;
    this->partyNumber = partyNumber;
    this->voteDestination = voteDestination;
    this->party = NULL;
}

const string& Candidate::get_balbox_name() const {
    return this->balboxName;
}

int Candidate::get_balbox_number() const {
    return this->balboxNumber;
}

int Candidate::get_nominal_votes() const {
    return this->nominalVotes;
}

const string& Candidate::get_situation() const {
    return this->situation;
}

int Candidate::get_party_number() const {
    return this->partyNumber;
}

const string& Candidate::get_vote_destination() const {
    return this->voteDestination;
}

PoliticParty* Candidate::get_party() const {
    return this->party;
}

void Candidate::set_party(PoliticParty* party) {
    this->party = party;
}

bool Candidate::elected() const {
    return this->get_situation() == "Eleito";
}

bool Candidate::valid_vote() const {
    return this->get_vote_destination() == "Válido";
}

ostream& operator<<(ostream& out, const Candidate& candidate) {
    out << candidate.get_name()<< endl;
    // out << candidate.get_gender() << endl;
    // out << formatDate(candidate.get_birth(), "%d/%m/%Y") << endl;
    // out << candidate.balboxName << endl;
    // out << candidate.balboxNumber << endl;
    // out << candidate.nominalVotes << endl;
    // out << candidate.situation << endl;
    // out << candidate.partyNumber << endl;
    // out << candidate.voteDestination << endl;

    return out;
}

bool comparator_candidates(const Candidate* m, const Candidate* n) {
    if(n->get_nominal_votes() == m->get_nominal_votes()) {
        time_t birth1 = m->get_birth();
        time_t birth2 = n->get_birth();

        if(birth1 < birth2) return true;
        else return false;
    
    } else {
        if(n->get_nominal_votes() < m->get_nominal_votes()) return false;
        else return true;
    }
}
